### ISSUE TITLE
Add support for async lock functions in lock_api

### ISSUE THREAD

**Issue by Lochlanna** [Amanieu/parking_lot#443]
With the advent of async traits in Rust, I believe it would be highly beneficial for lock_api to provide a trait and implementation that exposes asynchronous lock functions for Mutex and RwLock. Implementing versions with timeouts may be unnecessary since the futures can be cancelled by dropping them, inherently handling timeouts.

To integrate this, the new functionality would need to be conditionally compiled based on compiler support for async traits. Here's a rough outline of what I'm proposing:

```rust
#[cfg(has_async_fn_in_trait)]
pub unsafe trait RawMutexAsync: RawMutex {
    /// Locks this mutex asynchronously.
    fn lock_async(&self) -> impl Future<Output = ()> + Send;
}

#[cfg(has_async_fn_in_trait)]
impl<R: RawMutex + RawMutexAsync, T: ?Sized> Mutex<R, T> {
    pub async fn lock_async(&self) -> MutexGuard<'_, R, T> {
        RawMutexAsync::lock_async(&self.raw).await;
        // SAFETY: The lock is held, as required.
        unsafe { self.make_guard_unchecked() }
    }

    #[cfg(feature = "arc_lock")]
    pub async fn lock_async_arc(self: &Arc<Self>) -> ArcMutexGuard<R, T> {
        RawMutexAsync::lock_async(&self.raw).await;
        // SAFETY: The lock is held, as required.
        unsafe { self.make_arc_guard_unchecked() }
    }
}
```

A potential issue is the reliance on `RawMutex` being implemented as well. This approach wouldn't work for purely async locks. One possible workaround could be to split the `RawMutex` trait into two traits: one containing the shared items and another without them. However, this would be a significant change.

I'd love to hear your thoughts on this.

**Comment by Amanieu** [comment]
My position has always been that async locks need a completely different interface and implementation than normal lock, and therefore should be in a separate crate from parking_lot and lock_api.

**Comment by Lochlanna** [comment]
Makes sense. I'm trying to figure out the best way to setup my lock which is a hybrid that supports both sync and async at the same time. I'll probably just roll a custom API for it.
